server/eventlog/eventlog.go

```diff
@@ -23,41 +37,243 @@ func GetEventLogChunk(ctx context.Context, env environment.Env, req *elpb.GetEve
+		boundary = math.MaxUint16
+		step = math.MaxUint16 // decrements the value when added
+	}
+	// Fetch one chunk even if the minimum line count is 0
+	for lineCount := 0; (rsp.Chunk.ChunkId != "" || lineCount < int(req.MinLines)) && chunkIndex != boundary; chunkIndex += step {
+		buffer, err := c.ReadChunk(ctx, eventLogPath, chunkIndex)
+		if err != nil {
+			return nil, err
+		}
+		scanner := bufio.NewScanner(bytes.NewReader(buffer))
+		for scanner.Scan() {
+			if scanner.Err() != nil {
+				return nil, err
+			}
+			lineCount++
+		}
+		if readBackward {
+			rsp.Chunk.Buffer = append(buffer, rsp.Chunk.Buffer...)
+		} else {
+			rsp.Chunk.Buffer = append(rsp.Chunk.Buffer, buffer...)
+		}
+		if rsp.Chunk.ChunkId != chunkstore.ChunkIndexAsString(chunkIndex) {
+			// No longer fetching a single chunk, ChunkId cannot be meaningfully set.
+			rsp.Chunk.ChunkId = ""
+		}
 	}
-	if intChunkId > 0 {
-		rsp.PreviousChunkId = fmt.Sprintf("%04x", intChunkId-1)
+	if readBackward {
+		rsp.PreviousChunkId = chunkstore.ChunkIndexAsString(chunkIndex)
+	} else {
+		rsp.NextChunkId = chunkstore.ChunkIndexAsString(chunkIndex)
 	}
-	if intChunkId < math.MaxInt16 {
-		rsp.NextChunkId = fmt.Sprintf("%04x", intChunkId+1)
+
+	invalidChunkId := chunkstore.ChunkIndexAsString(math.MaxUint16)
+	if rsp.NextChunkId == invalidChunkId {
+		rsp.NextChunkId = ""
+	}
+	if rsp.PreviousChunkId == invalidChunkId {
+		rsp.PreviousChunkId = ""
 	}
 
-	print(rsp)
 	return rsp, nil
 }
+
+func NewEventLogWriter(ctx context.Context, b interfaces.Blobstore, invocationId string) *EventLogWriter {
+	return &EventLogWriter{
+		chunkstore.New(
+			b,
+			&chunkstore.ChunkstoreOptions{
+				WriteBlockSize:       defaultLogChunkSize,
+				WriteTimeoutDuration: defaultChunkTimeout,
+				NoSplitWrite:         true,
+			},
+		).Writer(ctx, GetEventLogPathFromInvocationId(invocationId)),
+	}
+}
+
+type EventLogWriter struct {
+	*chunkstore.ChunkstoreWriter
+}
+
+func (w *EventLogWriter) WriteLines(p []byte) (int, error) {
+	bytesWritten := 0
+	scanner := bufio.NewScanner(bytes.NewReader(p))
+	if scanner.Err() != nil {
+		return bytesWritten, scanner.Err()
+	}
+	line := scanner.Bytes()
+	for scanner.Scan() {
+		n, err := w.ChunkstoreWriter.Write(append(line, '\n'))
+		bytesWritten += n
+		if err != nil {
+			return bytesWritten, err
+		}
+		if scanner.Err() != nil {
+			return bytesWritten, scanner.Err()
+		}
+		line = scanner.Bytes()
+	}
+	if len(p) != 0 && p[len(p)-1] == '\n' {
+		line = append(line, '\n')
+	}
+	n, err := w.ChunkstoreWriter.Write(append(line, '\n'))
+	bytesWritten += n
+	return bytesWritten, err
+}
+
+// Split ANSI text by line, rewriting ANSI control sequences when necessary such
+// that any line printed individually is still formatted as it was in the
+// original buffer
+func (w *EventLogWriter) WriteLinesANSI(p []byte) (int, error) {
+	s := string(p)
+	if !ansi.HasEscapeCodes(s) {
+		return w.WriteLines(p)
+	}
+	remainingSegments, err := ansi.Parse(string(s))
```

Comment:
nit: add a comment explaining what a segment is?

Reply:
done.